velox/substrait/VeloxToSubstraitPlan.cpp

```diff
@@ -0,0 +1,158 @@
+  auto inputType = source->outputType();
+  filterRel->set_allocated_condition(
+      exprConvertor_.toSubstraitExpr(arena, filterCondition, inputType));
+
+  filterRel->mutable_common()->mutable_direct();
+}
+
+void VeloxToSubstraitPlanConvertor::toSubstrait(
+    google::protobuf::Arena& arena,
+    const std::shared_ptr<const ValuesNode>& valuesNode,
+    ::substrait::ReadRel* readRel) {
+  const auto& outputType = valuesNode->outputType();
+
+  ::substrait::ReadRel_VirtualTable* virtualTable =
+      readRel->mutable_virtual_table();
+
+  readRel->set_allocated_base_schema(
+      typeConvertor_.toSubstraitNamedStruct(arena, outputType));
+  // The row number of the input data.
+  int64_t numRows = valuesNode->values().size();
+
+  // There have multiple rows in the data and each row is a RowVectorPtr.
+  for (int64_t row = 0; row < numRows; ++row) {
+    // The row data.
+    ::substrait::Expression_Literal_Struct* litValue =
+        virtualTable->add_values();
+    const auto& rowValue = valuesNode->values().at(row);
+    // The column number of the row data.
+    int64_t numColumns = rowValue->childrenSize();
+
+    for (int64_t column = 0; column < numColumns; ++column) {
+      ::substrait::Expression_Literal* substraitField =
+          google::protobuf::Arena::CreateMessage<
+              ::substrait::Expression_Literal>(&arena);
+
+      VectorPtr children = rowValue->childAt(column);
+
+      substraitField->MergeFrom(*exprConvertor_.toSubstraitExpr(
+          arena, std::make_shared<ConstantTypedExpr>(children), litValue));
+    }
+  }
+  readRel->mutable_common()->mutable_direct();
+}
+
+void VeloxToSubstraitPlanConvertor::toSubstrait(
+    google::protobuf::Arena& arena,
+    const std::shared_ptr<const ProjectNode>& projectNode,
+    ::substrait::ProjectRel* projectRel) {
+  std::vector<std::shared_ptr<const ITypedExpr>> projections =
+      projectNode->projections();
+
+  std::vector<std::shared_ptr<const PlanNode>> sources = projectNode->sources();
+  // Check there only have one input.
+  VELOX_USER_CHECK_EQ(
+      1, sources.size(), "Project plan node must have extract one source.");
+  // The previous node.
+  std::shared_ptr<const PlanNode> source = sources[0];
+
+  // Process the source Node.
+  ::substrait::Rel* projectRelInput = projectRel->mutable_input();
+  toSubstrait(arena, source, projectRelInput);
+
+  // Remap the output.
+  ::substrait::RelCommon_Emit* projRelEmit =
+      projectRel->mutable_common()->mutable_emit();
+
+  int64_t projectionSize = projections.size();
+
+  auto inputType = source->outputType();
+  int64_t projectEmitReMapId = inputType->size();
+
+  for (int64_t i = 0; i < projectionSize; i++) {
+    std::shared_ptr<const ITypedExpr>& veloxExpr = projections.at(i);
+
+    projectRel->add_expressions()->MergeFrom(
+        *exprConvertor_.toSubstraitExpr(arena, veloxExpr, inputType));
+
+    // Add outputMapping for each expression.
+    projRelEmit->add_output_mapping(projectEmitReMapId);
```

Comment:
is projectEmitReMapId == inputType->size() + i ? If so, perhaps, we don't need projectEmitReMapId variable.